Create documentation (#59)
nf-core/datasync: docs/usage.md

Reviewer: Can you also document the behavior of the pipeline in all the edge test cases? 
Example: https://github.com/nf-core/datasync/issues/43

Perhaps we can have a short summary table with all the possible issues and what would happen in the pipeline or what the multiqc would report. 

Also document how to include or exclude certain files (we might have to link to external rclone arguments)

| `sample`       | Yes                 | Unique identifier used in task labels and output report names. It must not contain whitespace. Use a unique value for each row to prevent published report files from colliding.                                                        |
| `input`        | Yes                 | Source file or directory. This can be a local path, HTTP(S) URL, object-storage URL such as `s3://bucket/prefix`, or configured remote such as `source_s3:bucket/prefix` or `source_azure:container/prefix`. Whitespace is not allowed. |
| `output_path`  | Yes                 | Destination directory understood by rclone, such as `/archive/runs`, `s3://bucket/prefix`, or a configured `remote:path`. Whitespace is not allowed.                                                                                    |
| `checksum_md5` | One checksum column | MD5 sum file (`.csv` or `.tsv`) used to validate `input` before copying. Leave empty when using SHA-256 only.                                                                                                                           |

> | `checksum_sha` | One checksum column | SHA-256 sum file (`.csv` or `.tsv`) used to validate `input` before copying. Leave empty when using MD5 only.                                                                                                                           |

Reviewer 2: The format of these files should imho be described in here as well.

At least one checksum manifest is required on every row. If both are supplied, both validations run. Checksum files must use the format accepted by [`rclone checksum`](https://rclone.org/commands/rclone_checksum/): one hash and one path per line, with paths relative to the source root. Despite the permitted `.csv`/`.tsv` filename suffix, the contents are checksum-manifest text rather than a table with a header.

Example:

```csv title="samplesheet.csv"
sample,input,output_path,checksum_md5,checksum_sha
run_001,/data/run_001,s3://archive/runs,/data/checksums/run_001_md5.tsv
reference,https://example.org/reference.fa,/data/references,,/data/checksums/reference_sha256.tsv

> run_002,/data/run_002,archive:runs,/data/checksums/run_002_md5.tsv,/data/checksums/run_002_sha256.tsv

Reviewer 2: A description of form and format of the checksum files would be great. E.g. stating if paths are absolute or relative and of course what the required columns are.

```

> An [example samplesheet](../assets/samplesheet.csv) is included in the repository.

Reviewer 2: Can you also include an example checksum file?

Pipeline settings can be provided in a `yaml` or `json` file via `-params-file <file>`.

> [!WARNING]
> Do not use `-c <file>` to specify parameters as this will result in errors. Custom config files specified with `-c` must only be used for [tuning process resource specifications](https://nf-co.re/docs/running/run-pipelines#configuring-pipelines), other infrastructural tweaks (such as output directories), or module arguments (args).

The above pipeline run specified with a params file in yaml format:
Then provide that exact file to the pipeline:

```bash
nextflow run nf-core/datasync -profile docker -params-file params.yaml
nextflow run nf-core/datasync \
    -profile docker \
    --input samplesheet.csv \
    --outdir results \
    --rclone_config /secure/rclone.conf
```

with:
### S3 and S3-compatible storage

```yaml title="params.yaml"
input: './samplesheet.csv'
outdir: './results/'
genome: 'GRCh37'
<...>
```

You can also generate such `YAML`/`JSON` files via [nf-core/launch](https://nf-co.re/launch).
The main use case tested for nf-core/datasync is transferring files between S3 buckets. An S3 remote specifies the provider, region, and credentials. For example:

> ### Updating the pipeline

Reviewer: You can keep this original template section and text as it might be useful to users

When you run the above command, Nextflow automatically pulls the pipeline code from GitHub and stores it as a cached version. When running the pipeline after this, it will always use the cached version if available - even if the pipeline has been updated since. To make sure that you're running the latest version of the pipeline, make sure that you regularly update the cached version of the pipeline:

```bash
nextflow pull nf-core/datasync
```ini title="rclone.conf"
[s3]
type = s3
provider = AWS
access_key_id = YOUR_ACCESS_KEY_ID
secret_access_key = YOUR_SECRET_ACCESS_KEY
region = eu-central-1
```

> ### Reproducibility

Reviewer: You can keep this original template section and text as it might be useful to users

This version number will be logged in reports when you run the pipeline, so that you'll know what you used when you look back in the future. For example, at the bottom of the MultiQC reports.

To further assist in reproducibility, you can use share and reuse [parameter files](#running-the-pipeline) to repeat pipeline runs with the same settings without having to write out a command with every single parameter.

> [!TIP]
> If you wish to share such profile (such as upload as supplementary material for academic publications), make sure to NOT include cluster specific paths to files, nor institutional specific profiles.
The corresponding input values could be `source_s3:incoming/run_001` and `institutional_s3:project/run_002`. Provider-specific settings vary: consult the [rclone S3 documentation](https://rclone.org/s3/) and your storage provider's endpoint, region, addressing-style, and credential documentation rather than copying example values unchanged.

> ## Core Nextflow arguments

Reviewer: You can keep this original template section and text as it might be useful to users

### `-resume`

Specify this when restarting a pipeline. Nextflow will use cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously. For input to be considered the same, not only the names must be identical but the files' contents as well. For more info about this parameter, see [this blog post](https://www.nextflow.io/blog/2019/demystifying-nextflow-resume.html).

You can also supply a run name to resume a specific run: `-resume [run-name]`. Use the `nextflow log` command to show previous run names.

### `-c`
## Running the pipeline

Specify the path to a specific config file (this is a core Nextflow command). See the [nf-core website documentation](https://nf-co.re/usage/configuration) for more information.
A typical local-to-cloud run is:

## Custom configuration
```bash
nextflow run nf-core/datasync \
    -r <VERSION> \
    -profile docker \
    --input /data/samplesheet.csv \
    --outdir /data/datasync-results \
    --rclone_config /secure/rclone.conf
```

> ### Resource requests

Reviewer: Same with these sections

> The rclone processes use the `process_low` label. Configure executors and override CPU, memory, or time in a Nextflow config, for example:

Reviewer: ```suggestion
The `rclone` processes use the `process_low` label. Configure executors and override CPU, memory, or time in a Nextflow config, for example:
```

> Run with `-c resources.config`. Rclone derives its checker count from allocated CPUs, and the copy step uses roughly half that count (minimum one) for parallel transfers.

Reviewer: ```suggestion
Run with `-c resources.config`. `rclone` derives its checker count from allocated CPUs, and the copy step uses roughly half that count (minimum one) for parallel transfers.
```

## :warning: Please read this documentation on the nf-core website: [https://nf-co.re/datasync/usage](https://nf-co.re/datasync/usage)

> Pipeline parameter documentation is generated automatically from [`nextflow_schema.json`](../nextflow_schema.json). This page explains how to prepare a transfer and operate the pipeline.

## Prerequisites

Install Nextflow 25.10.4 or later and use a supported software profile. Docker or Singularity/Apptainer is recommended for reproducibility. Ensure that the account running Nextflow can read each source and checksum manifest and can write to every destination.

> For cloud or other authenticated rclone remotes, create an [rclone configuration](https://rclone.org/docs/) and pass it with `--rclone_config`. The configuration applies to remote **sources and destinations**. The pipeline has currently been tested for transfers between S3 buckets. Other rclone-supported layouts, such as Azure Blob Storage to S3 or transfers between S3-compatible providers, should be configured and validated against the upstream rclone documentation for each provider before use.

Reviewer: ```suggestion
For cloud or other authenticated rclone remotes, create an [rclone configuration](https://rclone.org/docs/) and pass it with `--rclone_config`. The configuration applies to remote **sources and destinations**. The pipeline has currently been tested for transfers between S3 buckets. Other rclone-supported layouts, such as Azure Blob Storage to S3 or transfers between S3-compatible providers, should be configured and validated against the upstream `rclone` documentation for each provider before use.
```

Example samplesheet:

```csv title="samplesheet.csv"
sample,input,output_path,checksum_md5,checksum_sha
run_001,/data/run_001,s3://archive/runs,/data/checksums/run_001_md5.tsv,
reference,https://example.org/reference.fa,/data/references,,/data/checksums/reference_sha256.tsv
run_002,/data/run_002,archive:runs,/data/checksums/run_002_md5.tsv,/data/checksums/run_002_sha256.tsv
```

For the `run_001` directory example, `/data/checksums/run_001_md5.tsv` could contain:

```text title="run_001_md5.tsv"
d41d8cd98f00b204e9800998ecf8427e  reads/sample_R1.fastq.gz
0cc175b9c0f1b6a831c399e269772661  reads/sample_R2.fastq.gz
900150983cd24fb0d6963f7d28e17f72  reports/qc_summary.txt
```

For a SHA-256 manifest, the same relative paths are used with SHA-256 hashes:

```text title="run_001_sha256.tsv"
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  reads/sample_R1.fastq.gz
ca978112ca1bbdcafac231b39a23dc4da786eff8147c4e72b9807785afee48bb  reads/sample_R2.fastq.gz
ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad  reports/qc_summary.txt
```

An [example samplesheet](../assets/samplesheet.csv) is included in the repository.

> ## Configuring rclone remotes

Reviewer: ```suggestion
## Configuring `rclone` remotes
```

> The file supplied with `--rclone_config` uses rclone's INI-style format. Each `[name]` section defines a remote, and samplesheet paths refer to it as `name:path`. The remote name is an arbitrary local label; it does not need to match the provider or bucket name. The pipeline's documented and tested configuration pattern is S3-to-S3 transfer. One file may contain several sections, so other cloud-to-cloud transfers can define both providers in the same file, but provider-specific options should be taken from the relevant rclone documentation rather than inferred from the S3 example:

Reviewer: ```suggestion
The file supplied with `--rclone_config` uses `rclone`'s INI-style format. 

Each `[name]` section defines a remote, and samplesheet paths refer to it as `name:path`. The remote name is an arbitrary local label; it does not need to match the provider or bucket name. 

> [!NOTE]
> The pipeline's documented and tested configuration pattern is S3-to-S3 transfer. 

One file may contain several sections, so other cloud-to-cloud transfers can define both providers in the same file, but provider-specific options should be taken from the relevant `rclone` documentation. This is an example using S3:
```

Reviewer: Can you add an URL pointing to the rclone config INI style format (or gneral rclone docs on how to prepare that file)?

```text
source_s3:incoming/run_001
archive_azure:research-archive/run_001
```

Create the file interactively where possible:

```bash
rclone config --config /secure/rclone.conf
rclone listremotes --config /secure/rclone.conf
```

Then provide that exact file to the pipeline:

```bash
nextflow run nf-core/datasync \
    -profile docker \
    --input samplesheet.csv \
    --outdir results \
    --rclone_config /secure/rclone.conf
```

### S3 and S3-compatible storage

The main use case tested for nf-core/datasync is transferring files between S3 buckets. An S3 remote specifies the provider, region, and credentials. For example:

```ini title="rclone.conf"
[s3]
type = s3
provider = AWS
access_key_id = YOUR_ACCESS_KEY_ID
secret_access_key = YOUR_SECRET_ACCESS_KEY
region = eu-central-1
```

> The corresponding input values could be `source_s3:incoming/run_001` and `institutional_s3:project/run_002`. Provider-specific settings vary: consult the [rclone S3 documentation](https://rclone.org/s3/) and your storage provider's endpoint, region, addressing-style, and credential documentation rather than copying example values unchanged.

Reviewer: ```suggestion
The corresponding input values could be `source_s3:incoming/run_001` and `institutional_s3:project/run_002`. Provider-specific settings vary: consult the [`rclone` S3 documentation](https://rclone.org/s3/) and your storage provider's endpoint, region, addressing-style, and credential documentation rather than copying example values unchanged.
```

> The pipeline also accepts an `s3://bucket/path` source or destination. In that form, ensure credentials and provider settings are available to both Nextflow and rclone in the execution environment. A named remote such as `source_s3:bucket/path` makes the selected configuration section explicit and is preferable when a config file contains multiple S3 providers.

Reviewer: ```suggestion
The pipeline also accepts an `s3://bucket/path` source or destination. In that form, ensure credentials and provider settings are available to both Nextflow and `rclone` in the execution environment. A named remote such as `source_s3:bucket/path` makes the selected configuration section explicit and is preferable when a config file contains multiple S3 providers.
```

## Destination layout

The pipeline preserves the source basename:

> - for a file source, rclone copies the file into `output_path`, and validation expects `output_path/<source filename>`;

Reviewer: ```suggestion
- for a file source, `rclone` copies the file into `output_path`, and validation expects `output_path/<source filename>`;
```

- for a directory source, the pipeline appends the source directory name, so `/data/run_001` with `output_path=/archive/runs` is copied and checked at `/archive/runs/run_001`.

> A trailing slash on `output_path` is removed before these paths are constructed. Ensure that a destination does not already contain unrelated files: post-copy validation uses `rclone check --one-way`, which checks that source content exists and matches at the destination while tolerating destination-only files.

Reviewer: ```suggestion
A trailing slash on `output_path` is removed before these paths are constructed. Ensure that a destination does not already contain unrelated files: post-copy validation uses `rclone check --one-way`, which checks that source content exists and matches at the destination while tolerating destination-only files. You can override this behavior by providing your own config file with external arguments for `rclone check`. 
```

You can also generate such `YAML`/`JSON` files via [nf-core/launch](https://nf-co.re/launch).

### Parameter files

Frequently reused settings can be stored in YAML or JSON and loaded with `-params-file`:

```yaml title="params.yaml"
input: /data/samplesheet.csv
outdir: /data/datasync-results
rclone_config: /secure/rclone.conf
multiqc_title: July archive transfer
```

```bash
nextflow run nf-core/datasync -r <VERSION> -profile docker -params-file params.yaml
```

Do not use `-c` for pipeline parameters. Use it only for Nextflow executor, resources, and other infrastructure configuration.

### Common integrity outcomes

> The workflow is designed to collect rclone reports even when rclone detects differences. The table below summarises common edge cases and how to interpret them in the published reports and MultiQC.

Reviewer: ```suggestion
The workflow is designed to collect `rclone` reports even when `rclone` detects differences. The table below summarises common edge cases and how to interpret them in the published reports and MultiQC.
```

| Destination file exists but content differs from source            | Post-copy `rclone check`                                         | `*` / mismatch                    | Re-copy or investigate concurrent source/destination changes.                                                                                       |
| Destination contains files absent from the source                  | Post-copy `rclone check`                                         | `+` / missing from source         | The post-copy check uses `--one-way`, so destination-only files are tolerated, but should still be reviewed for unexpected stale or unrelated data. |
| Dry-run execution                                                  | Copy step uses `--dry-run`; checksum and check reports still run | Depends on existing destination   | No transfer data is written. Post-copy reports describe whatever was already present at the destination.                                            |

### Including or excluding files

> Filter files by passing additional rclone filter flags to the relevant rclone module through a Nextflow configuration file. Rclone supports flags such as `--include`, `--exclude`, `--filter`, `--files-from`, and related rule files; see the [rclone filtering documentation](https://rclone.org/filtering/) for rule syntax and ordering.

Reviewer: ```suggestion
Filter files by passing additional `rclone` filter flags to the relevant rclone module through a Nextflow configuration file. `rclone` supports flags such as `--include`, `--exclude`, `--filter`, `--files-from`, and related rule files; see the [`rclone` filtering documentation](https://rclone.org/filtering/) for rule syntax and ordering.
```

For example, to copy and check only FASTQ files while excluding temporary files, create a small infrastructure config:

```groovy title="rclone_filters.config"
process {
    withName: 'RCLONE_COPY' {
        ext.args = {
            [
                '--log-level INFO',
                '--stats 30s',
                '--stats-one-line',
                '--stats-log-level INFO',
                '--s3-chunk-size 64M',
                '--no-check-certificate',
                params.rclone_dry_run ? '--dry-run' : '',
                '--include "*.fastq.gz"',
                '--include "*.fq.gz"',
                '--exclude "*.tmp"',
                '--exclude "*"'
            ].findAll { it }.join(' ')
        }
    }

    withName: 'RCLONE_CHECK' {
        ext.args = {
            [
                '--no-check-certificate',
                '--one-way',
                '--include "*.fastq.gz"',
                '--include "*.fq.gz"',
                '--exclude "*.tmp"',
                '--exclude "*"'
            ].join(' ')
        }
    }
}
```

> Run it with `-c rclone_filters.config` in addition to your normal profile and parameters. Because `ext.args` overrides module defaults, include the default rclone flags you still need when adding filters. Keep checksum manifests consistent with the same filtering rules: if a file is intentionally excluded from copy/check, remove it from the checksum manifest or generate a manifest for only the included files.

Reviewer: ```suggestion
Run it with `-c rclone_filters.config` in addition to your normal profile and parameters. Because `ext.args` overrides module defaults, include the default `rclone` flags you still need when adding filters. Keep checksum manifests consistent with the same filtering rules: if a file is intentionally excluded from copy/check, remove it from the checksum manifest or generate a manifest for only the included files.
```

## Understanding completion and integrity

> For each row, the pipeline first validates supplied checksum manifests, performs the copy, and then compares source and destination. Rclone comparison commands write status reports even when differences are found, allowing all results to be collected in MultiQC. Therefore, a successful Nextflow run means the workflow completed; it does **not by itself** prove every object matched. Review `multiqc/multiqc_report.html` and the reports under `rclone/`, especially lines marked `-`, `+`, `*`, or `!` (see [output documentation](output.md)).

Reviewer: ```suggestion
For each row, the pipeline first validates supplied checksum manifests, performs the copy, and then compares source and destination. `rclone` comparison commands write status reports even when differences are found, allowing all results to be collected in MultiQC. Therefore, a successful Nextflow run means the workflow completed; it does **not by itself** prove every object matched. Review `multiqc/multiqc_report.html` and the reports under `rclone/`, especially lines marked `-`, `+`, `*`, or `!` (see [output documentation](output.md)).
```
